File: src/test/scala/tlschannel/CipherTest.scala
package tlschannel

import org.junit.jupiter.api.{DynamicTest, Test, TestFactory, TestInstance}
import org.junit.jupiter.api.TestInstance.Lifecycle

import javax.net.ssl.SSLContext
import tlschannel.helpers.SslContextFactory
import tlschannel.helpers.SocketPairFactory
import tlschannel.helpers.Loops

import java.util
import scala.jdk.CollectionConverters._

@TestInstance(Lifecycle.PER_CLASS)
class CipherTest {

  val protocols = {
    val allProtocols = SSLContext.getDefault.getSupportedSSLParameters.getProtocols.toSeq
    allProtocols.filterNot(_ == "SSLv2Hello")
  }

  val dataSize = 200 * 1000

  // Test a half-duplex interaction, with renegotiation before reversing the direction of the flow (as in HTTP)
  @TestFactory
  def testHalfDuplexWithRenegotiation(): util.Collection[DynamicTest] = {
    println("testHalfDuplexWithRenegotiation():")
    val tests = for {
      protocol <- protocols
      ctxFactory = new SslContextFactory(protocol)
      cipher <- ctxFactory.allCiphers
    } yield {
      DynamicTest.dynamicTest(
        s"testHalfDuplexWithRenegotiation() - protocol: $protocol, cipher: $cipher",
        () => {
          val socketFactory = new SocketPairFactory(ctxFactory.defaultContext)
          val socketPair = socketFactory.nioNio(Some(cipher))
          Loops.halfDuplex(socketPair, dataSize, renegotiation = protocol <= "TLSv1.2")
          val actualProtocol = socketPair.client.tls.getSslEngine.getSession.getProtocol
          val p = s"$protocol ($actualProtocol)"
          println(f"$p%-18s $cipher%-50s")
        }
      )
    }
    tests.asJava
  }

  // Test a full-duplex interaction, without any renegotiation
  @TestFactory
  def testFullDuplex(): util.Collection[DynamicTest] = {
    val tests = for {
      protocol <- protocols
      ctxFactory = new SslContextFactory(protocol)
      cipher <- ctxFactory.allCiphers
    } yield {
      DynamicTest.dynamicTest(
        s"testFullDuplex() - protocol: $protocol, cipher: $cipher",
        () => {
          val socketFactory = new SocketPairFactory(ctxFactory.defaultContext)
          val socketPair = socketFactory.nioNio(Some(cipher))
          Loops.fullDuplex(socketPair, dataSize)
          val actualProtocol = socketPair.client.tls.getSslEngine.getSession.getProtocol
          val p = s"$protocol ($actualProtocol)"
          println(f"$p%-18s $cipher%-50s")
        }
      )
    }
    tests.asJava
  }

}

File: src/test/scala/tlschannel/CipherTest.java
package tlschannel;

import java.security.NoSuchAlgorithmException;
import java.util.*;
import java.util.stream.Collectors;
import javax.net.ssl.SSLContext;
import org.junit.jupiter.api.DynamicTest;
import org.junit.jupiter.api.TestFactory;
import org.junit.jupiter.api.TestInstance;
import org.junit.jupiter.api.TestInstance.Lifecycle;
import tlschannel.helpers.Loops;
import tlschannel.helpers.SocketGroups.SocketPair;
import tlschannel.helpers.SocketPairFactory;
import tlschannel.helpers.SslContextFactory;

@TestInstance(Lifecycle.PER_CLASS)
public class CipherTest {

    private final List<String> protocols;
    private final int dataSize = 200 * 1000;

    public CipherTest() {
        try {
            String[] allProtocols =
                    SSLContext.getDefault().getSupportedSSLParameters().getProtocols();
            protocols = Arrays.stream(allProtocols)
                    .filter(x -> !x.equals("SSLv2Hello"))
                    .collect(Collectors.toList());
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException();
        }
    }

    // Test a half-duplex interaction, with renegotiation before reversing the direction of the flow (as in HTTP)
    @TestFactory
    public Collection<DynamicTest> testHalfDuplexWithRenegotiation() {
        System.out.println("testHalfDuplexWithRenegotiation():");
        List<DynamicTest> tests = new ArrayList<>();
        for (String protocol : protocols) {
            SslContextFactory ctxFactory = new SslContextFactory(protocol);
            for (String cipher : ctxFactory.getAllCiphers()) {
                tests.add(DynamicTest.dynamicTest(
                        String.format("testHalfDuplexWithRenegotiation() - protocol: %s, cipher: %s", protocol, cipher),
                        () -> {
                            SocketPairFactory socketFactory = new SocketPairFactory(ctxFactory.defaultContext());
                            SocketPair socketPair = socketFactory.nioNio(
                                    Optional.of(cipher), Optional.empty(), true, false, Optional.empty());
                            Loops.halfDuplex(socketPair, dataSize, protocol.compareTo("TLSv1.2") < 0, false);
                            String actualProtocol = socketPair
                                    .client
                                    .tls
                                    .getSslEngine()
                                    .getSession()
                                    .getProtocol();
                            String p = String.format("%s (%s)", protocol, actualProtocol);
                            System.out.printf("%-18s %-50s\n", p, cipher);
                        }));
            }
        }
        return tests;
    }

    // Test a full-duplex interaction, without any renegotiation
    @TestFactory
    public Collection<DynamicTest> testFullDuplex() {
        List<DynamicTest> tests = new ArrayList<>();
        for (String protocol : protocols) {
            SslContextFactory ctxFactory = new SslContextFactory(protocol);
            for (String cipher : ctxFactory.getAllCiphers()) {
                tests.add(DynamicTest.dynamicTest(
                        String.format("testFullDuplex() - protocol: %s, cipher: %s", protocol, cipher), () -> {
                            SocketPairFactory socketFactory = new SocketPairFactory(ctxFactory.defaultContext());
                            SocketPair socketPair = socketFactory.nioNio(
                                    Optional.of(cipher), Optional.empty(), true, false, Optional.empty());
                            Loops.fullDuplex(socketPair, dataSize);
                            String actualProtocol = socketPair
                                    .client
                                    .tls
                                    .getSslEngine()
                                    .getSession()
                                    .getProtocol();
                            String p = String.format("%s (%s)", protocol, actualProtocol);
                            System.out.printf("%-18s %-50s\n", p, cipher);
                        }));
            }
        }
        return tests;
    }
}
